### PR TITLE
Multi-GCC BinaryBuilder Binary Dependencies for Julia v1.0

### LINKED ISSUE #38: Build failed to get libcoinblas
When building on Julia v1.0 and Arch Linux I get
```
ERROR: LoadError: LoadError: LibraryProduct(nothing, ["libcoinblas"], Symbol(""), "Prefix(/home/blegat/.julia/dev/Clp/deps/usr)") is not satisfied, cannot generate deps.jl!
Stacktrace:
 [1] #write_deps_file#134(::Bool, ::Function, ::String, ::Array{LibraryProduct,1}) at /home/blegat/.julia/packages/BinaryProvider/UTYxu/src/Products.jl:409
 [2] write_deps_file(::String, ::Array{LibraryProduct,1}) at /home/blegat/.julia/packages/BinaryProvider/UTYxu/src/Products.jl:396
 [3] top-level scope at none:0
 [4] include at ./sysimg.jl:29 [inlined]
 [5] include(::String) at ./loading.jl:1073
 [6] top-level scope at none:0
in expression starting at /home/blegat/.julia/dev/Clp/deps/build_COINBLASBuilder.v1.4.6.jl:45
in expression starting at /home/blegat/.julia/dev/Clp/deps/build.jl:46
```
However, I get
```
$ ls -R usr
usr:
downloads  lib  logs  manifests

usr/downloads:
COINBLASBuilder.v1.4.6.x86_64-linux-gnu.tar.gz  COINBLASBuilder.v1.4.6.x86_64-linux-gnu.tar.gz.sha256

usr/lib:
libcoinblas.a  libcoinblas.la  libcoinblas.so  libcoinblas.so.1  libcoinblas.so.1.4.6  pkgconfig

usr/lib/pkgconfig:
coinblas.pc

usr/logs:
COINBLASBuilder.log

usr/manifests:
COINBLASBuilder.v1.4.6.x86_64-linux-gnu.list
```

### PR DESCRIPTION
**Pull request by juan-pablo-vielma**
Provides binaries with new multi-gcc BinaryProvider that should work with official julia binaries and custom source-builds. In particular, it should fix #38 and JuliaOpt/Cbc.jl#56

To avoid future issues like JuliaOpt/Cbc.jl#56 (and https://github.com/JuliaOpt/SCS.jl/pull/107#issuecomment-423161755) the binaries are self contained:
- all dependencies are statically linked (except system libraries like libgfortran)
- coin-blas and coin-lapack are used instead of Julia provider versions as Arch ships a non-standard build that links to the system's openblas and BP cannot yet deal with this (https://github.com/JuliaOpt/SCS.jl/pull/107#issuecomment-423161755). 
- only Clp symbols are exported (symbols from dependencies like coin-lapack and coin-osi are not exported)

using the separate dynamic libraries for dependencies would not provide an advantage until there is Pkg3 support for binary dependencies (right now each COIN solver would download its own copy of each dependency, which could cause unexpected symbol clashes if the solvers are independently updated).

### CODE REVIEW

File: src/Clp.jl [hunk 1]
```diff
@@ -6,6 +6,7 @@ VERSION < v"0.7.0-beta2.199" && __precompile__()
 module Clp
 
 using Compat
+using Libdl
```
Comment (blegat) [inline src/Clp.jl]: Why is this needed ?

Comment (juan-pablo-vielma) [reply src/Clp.jl]: Tu suppress a warning about Clp not having Libdl in its dependencies (Libdl is used by BinaryProvider), e.g. : https://github.com/bicycle1885/CodecZlib.jl/pull/26

File: deps/build.jl [hunk 2]
```diff
@@ -1,58 +1,93 @@
...
+    Linux(:x86_64, libc=:glibc, compiler_abi=CompilerABI(:gcc7)) => ("$bin_prefix/ClpBuilder.v1.16.11.x86_64-linux-gnu-gcc7.tar.gz", "b955aadca6134586babe120c499aee6583715d4964075bf73bbc42216e5ba911"),
+    Linux(:x86_64, libc=:glibc, compiler_abi=CompilerABI(:gcc8)) => ("$bin_prefix/ClpBuilder.v1.16.11.x86_64-linux-gnu-gcc8.tar.gz", "9b9ae725558a642b6a03a327236d91b1fc8c2c99c3afed932a3921d27ef1cbf6"),
+    Windows(:x86_64, compiler_abi=CompilerABI(:gcc6)) => ("$bin_prefix/ClpBuilder.v1.16.11.x86_64-w64-mingw32-gcc6.tar.gz", "03252a30570e9740dcc5f1413dd3481a68e8c12c7373fc1ac2a58cec73f3ff6f"),
+    Windows(:x86_64, compiler_abi=CompilerABI(:gcc7)) => ("$bin_prefix/ClpBuilder.v1.16.11.x86_64-w64-mingw32-gcc7.tar.gz", "76078cbbf17456ea3b27145a3539a1f16c26ac261a13485068c8718d55aba6a3"),
+    Windows(:x86_64, compiler_abi=CompilerABI(:gcc8)) => ("$bin_prefix/ClpBuilder.v1.16.11.x86_64-w64-mingw32-gcc8.tar.gz", "705cd3fc170d1607899e9bcfe981c84ef5244a94187abbc479251c7f37b316e1"),
 )
+                    
+# To fix gcc4 bug in Windows
```
Comment (mlubin) [inline deps/build.jl]: There should also be a link to the issue here.

File: deps/build.jl [hunk 3]
```diff
@@ -1,58 +1,93 @@
...
-    # Alternatively, you could attempt to install from a separate provider,
-    # build from source or something even more ambitious here.
-    error("Your platform $(triplet(platform_key())) is not supported by this package!")
-end
 
+    # If we have a download, and we are unsatisfied (or the version we're
+    # trying to install is not itself installed) then load it up!
+    if unsatisfied || !isinstalled(dl_info...; prefix=prefix)
+        # Download and install binaries
+    # no dynamic dependencies until Pkg3 support for binaries
```
Comment (mlubin) [inline deps/build.jl]: Inconsistent indent level of the comments. Usually comments start at the indent level of the surrounding code, e.g.,
```julia
if foo
    f()
    # This is a comment.
end
```

**Comment by mlubin** [comment]
@juan-pablo-vielma note that I pushed changes to your branch, so you'll need to pull from your own branch before making additional commits.

**Comment by blegat** [comment]
Just tested the PR on ArchLinux and it works

File: deps/build.jl [hunk 4]
```diff
@@ -1,58 +1,94 @@
...
+using BinaryProvider # requires BinaryProvider 0.3.0 or later
 
 # Parse some basic command-line arguments
 const verbose = "--verbose" in ARGS
 const prefix = Prefix(get([a for a in ARGS if a != "--verbose"], 1, joinpath(@__DIR__, "usr")))
 products = [
-    LibraryProduct(prefix, String["libOsiClp"], :libOsiClp),
-    LibraryProduct(prefix, String["libClp"], :libClp),
-    LibraryProduct(prefix, String["libClpSolver"], :libClpSolver),
+    #LibraryProduct(prefix, ["libOsiClp"], :libOsiClp),
```
Comment (mlubin) [inline deps/build.jl]: Style nit: Remove commented code or explain why it's commented and when it should be uncomented.

Comment (juan-pablo-vielma) [reply deps/build.jl]: deleted. We may want it back on the BinaryBuilder side eventually, but not on the BinaryProvider side.

**Pull request merged**